### PR TITLE
Add CRITIC agent integration

### PR DESCRIPTION
**Pull request by nerdai**
# Description

This PR adds a new package, namely `llama-index-agent-introspective`. This package introduces `IntrospectiveAgent`'s that perform tasks while utilizing the "reflection" agentic pattern. Two reflection agents that differ by their reflection mechanisms are also supplied in this package. Thus three main classes are introduced:

1. `IntrospectiveAgentWorker`
2. `ToolInteractiveReflectionAgentWorker`
3. `SelfReflectionAgentWorker` (adapted and ported over from #13089 by @jerryjliu)

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New package
- [x] This change requires a documentation update — notebooks have been added and module guides updated.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    async def _arun_step(
+        self,
+        step: TaskStep,
+        task: Task,
+    ) -> TaskStepOutput:
+        """Run step."""
+        if self.verbose:
+            print(f"> Running step {step.step_id} for task {task.task_id}.\n")
+
+        # run CRITIC reflection
```
Comment (nerdai) [inline llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py:85]: Thinking here is that a `step` includes an entire CRITIC cycle of `correct` and `critique` sub steps

File: llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        task.extra_state["new_memory"].put(
+            ChatMessage(content=parsed_correct_response), role=MessageRole.ASSISTANT
+        )
+        new_steps = [
+            TaskStep(
+                task_id=task.task_id,
+                step_id=str(uuid.uuid4()),
+                input=task.input,
+                step_state={
+                    "prev_reasoning": parsed_correct_response,
```
Comment (nerdai) [inline llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py:130]: I'm not sure if I'm using this correctly, but it seems like this is just some intermediate step state that I can use to carry out the reasoning loop?

Comment (logan-markewich) [reply llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py:130]: It's basically state attached to the step. So on the next step, you could access this field on the step object

File: llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py (hunk 1 above)
Comment (logan-markewich) [reply llama-index-integrations/agent/llama-index-agent-critic/llama_index/agent/critic/step.py:85]: I think that makes sense

File: llama-index-core/llama_index/core/agent/introspective/step.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            + task.memory.get()
+            + task.extra_state["main"]["memory"].get_all()
+            + task.extra_state["reflection"]["memory"].get_all()
+        )
+
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        # run main agent
+        main_agent = self._main_agent_worker.as_agent()
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/agent/introspective/step.py:177]: by calling `as_agent()` every time, we are starting with a fresh memory -- not sure if intentional or not

File: llama-index-core/llama_index/core/agent/introspective/step.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        main_agent_response = main_agent.chat(
+            task.input
+        )  # or should i use step.input here?
+        task.extra_state["main"]["sources"] = main_agent_response.sources
+        task.extra_state["main"]["memory"] = main_agent.memory
+        print(f"MAIN AGENT MEMORY: {main_agent.memory}", flush=True)
+
+        # run reflective agent
+        reflective_agent = self._reflective_agent_worker.as_agent()
+        reflective_agent_response = reflective_agent.chat(main_agent_response.response)
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/agent/introspective/step.py:187]: the only thing the reflective agent has for context here is the response from the first agent. Shouldn't it also have some context on what the original query was too?

File: llama-index-core/llama_index/core/agent/introspective/step.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        # run main agent
+        main_agent = self._main_agent_worker.as_agent()
+        main_agent_response = main_agent.chat(
+            task.input
+        )  # or should i use step.input here?
+        task.extra_state["main"]["sources"] = main_agent_response.sources
+        task.extra_state["main"]["memory"] = main_agent.memory
```
[the remark this answers is not in the document]
Comment (nerdai) [reply llama-index-core/llama_index/core/agent/introspective/step.py:182]: Oh this is a `BaseAgentRunner` so it should have memory right? It maybe empty `ChatMemoryBuffer`, but the attr should exist if that's what you're wondering about?

File: llama-index-core/llama_index/core/agent/introspective/step.py (hunk 4 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/agent/introspective/step.py:187]: Yeah, that's a good point!

File: llama-index-core/llama_index/core/agent/introspective/step.py (hunk 3 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/agent/introspective/step.py:177]: Right, unless I pass in `chat_history` or `memory` as part of `kwargs`?

File: llama-index-core/llama_index/core/agent/introspective/step.py (hunk 4 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/agent/introspective/step.py]: This is the only one that I haven't resolved yet. I do pass the chat history to the ToolInteractiveReflectionAgent, but not to the critique agent which gets the task of reflection with tools.

To do this would require a bit of re-jigging on the prompts for the CritiqueAgent and then ofc the passing of the chat history, which can be done as prompt str.

I think it would be okay to do this in a future release of this package, in order to get this package out sooner. Thoughts?

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        self._verbose = verbose
+
+        # define _reflection and _correction programs
+        try:
+            from llama_index.program.openai import OpenAIPydanticProgram
+        except ImportError:
+            raise ImportError(
+                "Missing OpenAIPydanticProgram. Please run `pip install llama-index-program-openai`."
+            )
+        self._reflection_program = OpenAIPydanticProgram.from_defaults(
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:160]: Instead of using programs (and locking to `OpenAIPydanticProgram` -- we can use `llm.structured_predict()` which picks the program that matches the LLM being used

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+            **kwargs,
+        )
+
+    def initialize_step(self, task: Task, **kwargs: Any) -> TaskStep:
+        """Initialize step from task."""
+        # temporary memory for new messages
+        new_memory = ChatMemoryBuffer.from_defaults()
+
+        # put current history in new memory
+        messages = task.memory.get()
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:210]: I don't think there's a perfect solution to this, but something to be aware of

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+        return ChatMessage.from_str(correct_response_str, role="assistant")
+
+    @dispatcher.span
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        state = step.step_state
+        state["count"] += 1
+
+        messages = task.extra_state["new_memory"].get()
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:268]: Does this include the initial task input? I might have missed where that gets added to memory

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+        if self.stopping_callable:
+            is_done = self.stopping_callable(critique_str=critique_response.response)
+
+        critique_msg = ChatMessage(
+            role=MessageRole.USER, content=critique_response.response
+        )
+        task.extra_state["new_memory"].put(critique_msg)
+
+        # correct
+        if is_done:
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py:353]: If `self.stopping_callable` is none, is this just never done?

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+            + task.memory.get()
+            + task.extra_state["main"]["memory"].get_all()
+            + task.extra_state["reflection"]["memory"].get_all()
+        )
+
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        # run main agent
+        if self._main_agent_worker is not None:
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py:143]: interesting, so you don't actually need a main agent worker 👀

File: llama-index-integrations/agent/llama-index-agent-introspective/tests/test_agent_critic.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+from llama_index.agent.introspective import (
+    IntrospectiveAgentWorker,
+    SelfReflectionAgentWorker,
+    ToolInteractiveReflectionAgentWorker,
+)
+from llama_index.core.agent.types import BaseAgentWorker
+
+
+def test_classes():
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-introspective/tests/test_agent_critic.py:9]: I wonder if it's worth writing one or two actual tests? You could copy what I did in `llama-index-core/tests/agent/runner/test_planner.py` -- just mock the LLM response depending on the input, and test that internal calls are working properly.

Not a requirement for this PR for now though

**Review by logan-markewich — COMMENTED** [review]
This looks good to me! Just a few comments/poossible edge cases.

Note for myself though -- need to do a better job of thinking about what goes in an integration vs. core 😅

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:210]: Thanks! I wonder if we should retain full memory and only use truncation when its about to be passed to an LLM.

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py (hunk 6 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:160]: Oh cool. Let me try using that instead for both reflection classes. Less params for a user to have to supply then. :)

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py (hunk 9 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py:353]: oh in that case it will stop after `max_iterations` has been reached. That gets implemented when building the `TaskStepOutput` with `is_last` field.

```python
        return TaskStepOutput(
            output=agent_response,
            task_step=step,
            is_last=is_done | (self.max_iterations == state["count"]),
            next_steps=new_steps,
        )
```

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py (hunk 10 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py:143]: Yeah, in this case the initial response is assumed to be provided within the user's input. In other words, we'll take the user input and just have the assistant repeat the input to "fictitiously" say that it generated a response. Thereafter reflection/correction cycles occur.

Importantly, the main agent worker is only responsible for generating the initial response. After that its not used again as the reflection agent takes care of reflection/correction.

(I have written this out in docs and README too so hopefully should be clear to the users of the package)

File: llama-index-integrations/agent/llama-index-agent-introspective/tests/test_agent_critic.py (hunk 11 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/tests/test_agent_critic.py:9]: Yeah! I can do that. Thanks, I'll certainly tackle this not too long (1-2 days max) after this gets merged.

**Comment by nerdai** [comment]
> This looks good to me! Just a few comments/poossible edge cases.
> 
> Note for myself though -- need to do a better job of thinking about what goes in an integration vs. core 😅

Yeah, I had initially thought the `IntrospectiveAgent` should be in core as reflection is a key pattern. But I opted to keep it contained in this package with reflection agents cause it seemed like a better package together than not.

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py (hunk 8 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:268]: yes it should, since the `introspective agent` first adds the chat history from the main agent which includews the initial task input to the reflective agent's memory.

https://github.com/run-llama/llama_index/blob/0d05f9090211a0c51a3796f052118fc5f720fc46/llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py#L162

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py (hunk 6 above)
Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py]: this is done!

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py:190]: I think how the react agent handles it is doing task_memory.get() + new_memory.get() -- but that also seems risky/not ideal.

Work for a future PR perhaps :)

File: llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py (hunk 9 above)
Comment (logan-markewich) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py:348]: Hmm, interesting. I wonder if that's ideal or not. I would expect a combination of the main agent being is_done and the reflection agent being is_done to control when we return 🤔 Maybe work for a future PR

Comment (nerdai) [reply llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py:348]: oh interesting yea, at this point it always go to reflection and this `is_done` is about when reflection/correction phases stop.

**Comment by nerdai** [comment]
alrighty @logan-markewich -- as discussed offline, I've cleaned up the memory for introspective agent removing the chat histories of the inner agents that it delegates tasks to. The final message in the history now lines up more nicely with the final response.

image
